cpp/src/arrow/python/flight.cc

```diff
@@ -130,6 +130,22 @@ Status PyFlightDataStream::Next(arrow::flight::FlightPayload* payload) {
   return stream_->Next(payload);
 }
 
+PyGeneratorFlightDataStream::PyGeneratorFlightDataStream(
+    PyObject* generator, std::shared_ptr<arrow::Schema> schema,
+    PyGeneratorFlightDataStreamCallback callback)
+    : schema_(schema), callback_(callback) {
+  Py_INCREF(generator);
+  generator_.reset(generator);
+}
+
+std::shared_ptr<arrow::Schema> PyGeneratorFlightDataStream::schema() { return schema_; }
+
+Status PyGeneratorFlightDataStream::Next(arrow::flight::FlightPayload* payload) {
+  PyAcquireGIL lock;
```

Comment:
For me it seems that `Next` is called while the GIL is held. This might end in deadlocks.

Reply:
I might be mistaken, but is it not safe to call `PyGILState_Ensure` even while you hold the GIL?

Reply:
This will call `PyGILState_Release` and thus release the GIL even if it did acquire it.

Reply:
But doesn't that only release the GIL if it was not held prior to the matching `PyGILState_Ensure` call (based on the [PEP that proposed it](https://www.python.org/dev/peps/pep-0311/#implementation) and the [current implementation](https://github.com/python/cpython/blob/8479a3426eb7d1840473f7788e639954363ed37e/Python/pystate.c#L1184))? (I'd rather be defensive and make sure the GIL is acquired even if so far we know it already is.)

Reply:
@wesm @pitrou Can you add comments to `PyAcquireGIL` that explains how it works in this case? I feel a bit confused now and that will probably happen to other people in future again.

Reply:
According to the Python docs

> Ensure that the current thread is ready to call the Python C API regardless of the current state of Python, or of the global interpreter lock. This may be called as many times as desired by a thread as long as each call is matched with a call to PyGILState_Release(). In general, other thread-related APIs may be used between PyGILState_Ensure() and PyGILState_Release() calls as long as the thread state is restored to its previous state before the Release(). For example, normal usage of the Py_BEGIN_ALLOW_THREADS and Py_END_ALLOW_THREADS macros is acceptable.